### ISSUE TITLE
Show staging status

### ISSUE THREAD

**Issue by mcmillion** [jisaacks/GitGutter#118]
Would it be possible to show an indicator for lines of files that have been staged?

**Comment by Adarma** [comment]
@jisaacks 
I want to add the colours to my theme, but notice some inconsistencies between the scopes you mention:

- scope: markup.inserted.git_gutter
- scope: markup.deleted.git_gutter
- scope: markup.modified.git_gutter
- scope: markup.inserted.git_gutter.staged
- scope: markup.deleted.git_gutter.unstaged
- scope: markup.modified.git_gutter.staged_unstaged

... and the scopes in DeepBlueSee:

- scope: markup.deleted.git_gutter, markup.deleted.git_gutter.staged
- scope: markup.inserted.git_gutter, markup.inserted.git_gutter.staged
- scope: markup.changed.git_gutter, markup.changed.git_gutter.staged
- scope: markup.deleted.git_gutter.unstaged
- scope: markup.inserted.git_gutter.unstaged
- scope: markup.changed.git_gutter.unstaged
- scope: markup.ignored.git_gutter
- scope: markup.untracked.git_gutter

Some Questions:

Should it be ´markup.modified...´ or ´markup.changed...´ or do they do different things?
Is ´markup.modified.git_gutter.staged_unstaged´ used?
How is ´markup.ignored.git_gutter´ used?
how is ´markup.untracked.git_gutter´ used?

Is this released yet?

**Comment by RafalSkolasinski** [comment]
Hmm... I am not sure if this should be separate issue or comment here... I was checking possible options for ``compare_against`` and I think it's missing comparing against staged area. In the way that showed changes are consistent with ``git status`` output.

**Comment by deathaxe** [comment]
I like this idea as it should by quite easy to realize and seems straight forward. I was thinking about the approach suggested by PR #136 to combine diffs from staging area and HEAD but found it quite heavy. 

I think I'll pick up this idea.

**Comment by deathaxe** [comment]
### An idea how to handle _Compare against staging area_

I find selecting the _Compare against staging area_  manually possibly not to be so user friendly and a bit tricky on the other hand.

1. From user's view I would prefere to automatically compare against the staged file if exists.
2. From the developer's view it is a bit tricky to set the compare target manually, because I can't compare against an empty staging area. Automatically switching compare target back to HEAD if a view does not show staged file might confuse users. The compare target is set per repository. Not all open views may show a staged file. If user changes views ... what to do?

So here is an idea:

GitGutter compares against the HEAD by default, but can compare the view against the current branch, too.

When comparing against...

1. HEAD: The view is automatically compared against the staging area if not empty.
2. current branch: The view is always compared against the latest commit.

These two options would allow comparing against both staging area and latest commit without any additional command.

So all change markers will disapear as soon as a file gets staged. The staging status is shown in the status bar. Any further changes will start to add new gutter icons and the status bar text displays _staged and modified_.

**Comment by jisaacks** [comment]
@deathaxe I am not sure I follow. What is the difference between HEAD and the current branch? Don't the both point to the same commit?

**Comment by deathaxe** [comment]
Yes this is the point. The question is, how and when to enable people to compare against the staging area.

One option would be to add a `Compare against staging area` command as suggested by @RafalSkolasinski. The user would need to activly select the staging area to compare against by keybinding, command pallet or setting.

But if a view's file is not in the staging area, GitGutter needs to fallback to the HEAD anyway to have something to compare against. This can happen in two ways:

1. Keep user selected compare target and simply compare against the HEAD.
2. Reset user selected compare target from 'staged' to 'HEAD'.

Option 2 interferes with the feature of selecting a compare target per repository. Example: 2 tabs showing files of the same repo are open. 1 staged, 1 modified. User has the staged view on top editing something while compared against staging area. Now he switches to second (modified) tab. The compare target is reset to HEAD as not contained in the staging area. Switching back to the staged view, would now no longer compare against staging area as HEAD is the new target. -> No-Go.

Option 1 is the better idea, as it would automatically compare against HEAD for unstaged files and compare against staging area if a file is staged. The new command would be named like `Compare against HEAD or staging area` then.

So the question is: Do I need to add this explicit command and setting value for it or can I automatically compare against staging area, if HEAD is the selected compare target? 

Doing so would extend the meaning of HEAD to include a staged file if available automatically without user interaction. If a user still wants to compare against the latest committed file the `Compare against branch` command could be used to select the current branch, which would not include staging area.

**Comment by Adarma** [comment]
I think you are going down the rabbit hole with this comparing to the staging area stuff.

The idea is to 'simply' compare to the HEAD as usual, but have different icons to indicate if the hunk is staged or not.

[I do realize this is not actually simple to implement...]

**Comment by deathaxe** [comment]
This is what I wanted to avoid because this requires to read both the staged and commited file, compare both to the view and merge results. This costs much more effort and time on large files. Even worse, a new scope is needed for a new set of icons, which will then need to be added to all color schemes to show up correctly. This is the approach from PR #136.

Just switching between staged and commited files would just need to add one further method to read the staged file content. Nothing else need to be changed. The staging are is handled like any other compare target.

**Comment by jisaacks** [comment]
> The idea is to 'simply' compare to the HEAD as usual, but have different icons to indicate if the hunk is staged or not.

This is precisely what I was attempting to solve in my branch. I was getting results but its a very complex problem. Probably more complex than all of the rest of git gutter. And it's an edge case. Not worth the complexity in my opinion. Although it would be cool.

**Comment by Adarma** [comment]
**Was there any progress on this?** I see there is still a "staged" branch.
**Does compare against branch now compare to the staging area if staged?**

I'm upgrading a colour scheme to the "sublime-color-scheme" format and have these scopes:

- markup.deleted.git_gutter, markup.deleted.git_gutter.staged
- markup.inserted.git_gutter, markup.inserted.git_gutter.staged
- markup.changed.git_gutter, markup.changed.git_gutter.staged
- markup.deleted.git_gutter.unstaged
- markup.inserted.git_gutter.unstaged
- markup.changed.git_gutter.unstaged
- markup.ignored.git_gutter
- markup.untracked.git_gutter

**Are all these scopes still used?**

The new SublimeLinter4 sets colours via its settings file rather than custom scopes defined in the color scheme. It has this setting to set the error colour:

- "scope": "region.redish markup.error.sublime_linter"

**Any plan to set colours a similar way?**

**Comment by deathaxe** [comment]
1. The `.staged` and `.unstaged` scopes were never used in any GitGutter release.
2. The `staged` branch is quite outdated and contains some unsolved issues in ways how to handle regions which were changed in both the staging area and the worktree.
3. I tried to create a way to support the index as normal compare target such as any other branch/tag/commit in the `status_porcelain2` branch, but it broke support for older git versions (<2.10) and introduces a couple of issues and complexities with all the existing "Compare against functions".  Therefore I did not progress on this path. The most tricky thing is to track the index for changes to avoid checking out its content after each keypress. I found a possible way, but it would require some bigger code changes I simple did not find the time/power for.

4. I know about the `region.redish` and co, but I find them too blurry. The correct scope for such kinds of regions is simply `markup.`. Therefore it is not planed to support the region. ... stuff.

**Comment by r-stein** [comment]
If you use the scope `region.redish markup.deleted.git_gutter` it will only use the color of `region.redish` if the colorscheme does not define a color for `markup` or `markup.deleted` or `markup.deleted.git_gutter`, imo it combines the best of both world. You always have color and it is not everything white, but a colorscheme can overwrite that color.

It looks like it would need a refactoring, but except from that I don't really see disadvantages. Nonetheless I am not sure whether we should expose that to the settings.
